### PR TITLE
Implement outer join and friends

### PR DESCRIPTION
**Pull request by marcelogomez**
This PR implements all of the methods described [this comment](https://github.com/jonhoo/hashbag/pull/6#issuecomment-1162596000) on #6 :
- `outer_join`: Takes in a reference to a second `HashBag` and returns an iterator that yields the union of the key sets for both `HashBag`s (ie. `keyset(self) ∪ keyset(other)`). Each key is yielded along with its corresponding counts on each `HashBag`.
  - The rest of the functions added in this PR as well as `difference` introduced in #6 can be implemented using this function
- `not_in`: Takes in a reference to a second `HashBag` and returns an iterator that yields the set difference between the key set of the first `HashBag` and the second one (i.e. `keyset(self) \ keyset(other)`). Each key is yielded along with its corresponding count on the first `HashBag`.
- `signed_difference`: Takes in a reference to a second `HashBag` and returns an iterator that yields the union of the keys sets for both `HashBag`s (ie. `keyset(self) ∪ keyset(other`). Each key is yielded along with the difference between its count on the first and second `HashBag`s.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -1256,4 +1342,95 @@ mod tests {
...
+        this: &[usize],
+        other: &[usize],
+        expected_entries: &[(&usize, usize, usize)],
+    ) {
+        let this_hashbag: HashBag<_> = this.iter().cloned().collect();
+        let other_hashbag: HashBag<_> = other.iter().cloned().collect();
+        let expected: HashSet<_> = HashSet::from_iter(expected_entries.iter().cloned());
+        let actual: HashSet<_> = this_hashbag.outer_join(&other_hashbag).collect();
+        assert_eq!(expected, actual);
+    }
```
Comment (marcelogomez) [inline src/lib.rs:1409]: Would appreciate some ideas on how to avoid some code duplication between all these `do_test_<fn>` methods. They're pretty similar, but slightly different, so far I haven't been able to come up with a good way of simplifying this.

I tried to write a generic method that takes `IntoIter`s for the inputs and expected values and a `FnOnce` that transform both inputs into the `actual` value we want to do `assert_eq!` on but that got ugly pretty quickly :/

**Review by jonhoo — CHANGES_REQUESTED** [review]
This looks great, thanks for taking it on! The only thing I'm missing is a test that checks that `outer_join` indeed returns all the elements of `self` before any elements that are exclusively in `other`. We promise that in the docs, so we should have a test that the implementation meets that promise.

File: src/lib.rs (hunk 1 above)
Comment (jonhoo) [reply src/lib.rs:1409]: I think it's totally fine to have separate helpers here. Over-generalizing can make the testing code itself hard to follow :)

**Comment by marcelogomez** [comment]
> This looks great, thanks for taking it on! The only thing I'm missing is a test that checks that `outer_join` indeed returns all the elements of `self` before any elements that are exclusively in `other`. We promise that in the docs, so we should have a test that the implementation meets that promise.

Thanks for taking a look! :D

Yeah, will add that test! However, that's not what I meant when I wrote:

> The iterator also yields the respective counts in `self` and `other` in that order.

The order I was referring was the the order of the values in the tuple that the iterator yields (i.e. it's `(key, self_count, other_count)`. Will change the documentation to make that clearer

**Comment by jonhoo** [comment]
> The order I was referring was the the order of the values in the tuple that the iterator yields (i.e. it's `(key, self_count, other_count)`. Will change the documentation to make that clearer

Ah, I see! Yeah, I think we want to make both promises in the docs (and check in tests) — that the first count in the tuple is from `self`, and that all elements not in `other` are yielded first.

**Comment by marcelogomez** [comment]
> Ah, I see! Yeah, I think we want to make both promises in the docs (and check in tests) — that the first count in the tuple is from `self`, and that all elements not in `other` are yielded first.

The tests for `outer_join` assert on the exact values yielded by the iterator so I think the ordering within the tuple is already covered  by tests :) I just added new tests for the ordering between yielded elements

**Review by jonhoo — APPROVED** [review]
Looks great, thanks!

**Comment by jonhoo** [comment]
Published as 0.1.7 :tada:

File: src/lib.rs [hunk 2]
```diff
@@ -724,14 +758,64 @@ where
...
+    /// let expected: HashSet<_> = HashSet::from_iter([(&1, 1), (&2, 0), (&3, 1), (&4, -2)]);
+    /// let actual: HashSet<_> = a.signed_difference(&b).collect();
+    /// assert_eq!(expected, actual);
+    /// ```
+    pub fn signed_difference<'a>(
+        &'a self,
+        other: &'a HashBag<T, S>,
+    ) -> impl Iterator<Item = (&'a T, isize)> {
+        self.outer_join(other)
+            .map(|(x, self_count, other_count)| (x, self_count as isize - other_count as isize))
```
Comment (Dr-Emann) [inline src/lib.rs:792]: Should we document/change the behavior for differences which can't fit in an `isize`?

```rust
let mut a: HashBag<_> = HashBag::new();
let mut b: HashBag<_> = HashBag::new();
a.insert_many(1, usize::MAX);
b.insert_many(2, usize::MAX - 10);
eprintln!("{:?}", a.signed_difference(&b).collect::<Vec<_>>());
```

currently outputs: `[(1, -1), (2, 11)]`. (even the signs are incorrect: `a` _does_ contain 1, so the value should be positive)

Comment (jonhoo) [reply src/lib.rs:792]: Oh, good call, we should definitely document that!

Comment (Dr-Emann) [reply src/lib.rs:792]: See #11
